### PR TITLE
442 fail2ban

### PR DESCRIPTION
**Pull request by Stamper-repo**
Proposal for: https://github.com/eikek/sharry/issues/442

Note to self:

    Authentication failures for logging in are logged
    Authentication failures for using wrong password on share is missing

### CODE REVIEW

File: modules/restserver/src/main/scala/sharry/restserver/routes/LoginRoutes.scala [hunk 1]
```diff
@@ -147,6 +148,7 @@ object LoginRoutes {
           ).map(_.addCookie(cd.asCookie(getBaseUrl(cfg, req))))
         } yield resp
       case _ =>
+        logger.info(s"Authentication attempt failure for username ${accountName} from ip ${req.from.map(_.getHostAddress).getOrElse("Unknown ip")}")
```
Comment (eikek) [inline modules/restserver/src/main/scala/sharry/restserver/routes/LoginRoutes.scala:151]: To get the clients ip/hostname, there is something `ClientRequestInfo.getHostname` which might be useful here (it  checks the usual headers but might need some adjustments for this use-case)

Comment (Stamper-repo) [reply modules/restserver/src/main/scala/sharry/restserver/routes/LoginRoutes.scala:151]: Cheers for the tip! The `req.from` from http4s checks the relevant headers and `getHostAddress` extracts just the ip address. I tested it successfully locally with an nginx setup. I also just pushed some commits with example configs.

Would this implementation fit for you?

**Comment by Stamper-repo** [comment]
@eikek The `sbt fix` command is giving me some elm errors after pulling the newest master:
```
[sebastiaan@localhost sharry]$ sbt fix
[info] welcome to sbt 1.5.2 (Oracle Corporation Java 1.8.0_144)
[info] loading settings for project global-plugins from idea.sbt ...
[info] loading global plugins from /home/sebastiaan/.sbt/1.0/plugins
[info] loading settings for project sharry-build from plugins.sbt ...
[info] loading project definition from /home/sebastiaan/repositories/sharry/project
[info] loading settings for project root from build.sbt,version.sbt ...
[info] set current project to sharry-root (in build file:/home/sebastiaan/repositories/sharry/)
[info] Compile elm files ...
[info] Compiling css stylesheets…
[info] Copy /home/sebastiaan/repositories/sharry/modules/webapp/src/main/webjar/img/logo-dark.svg -> /home/sebastiaan/repositories/sharry/modules/webapp/target/scala-2.13/resource_managed/main/META-INF/resources/webjars/sharry-webapp/1.8.0-SNAPSHOT/img/logo-dark.svg
[info] Copy /home/sebastiaan/repositories/sharry/modules/webapp/src/main/webjar/img/icon.svg -> /home/sebastiaan/repositories/sharry/modules/webapp/target/scala-2.13/resource_managed/main/META-INF/resources/webjars/sharry-webapp/1.8.0-SNAPSHOT/img/icon.svg
[info] Copy /home/sebastiaan/repositories/sharry/modules/webapp/src/main/webjar/img/logo.png -> /home/sebastiaan/repositories/sharry/modules/webapp/target/scala-2.13/resource_managed/main/META-INF/resources/webjars/sharry-webapp/1.8.0-SNAPSHOT/img/logo.png
[info] Copy /home/sebastiaan/repositories/sharry/modules/webapp/src/main/webjar/img/icon-dark.svg -> /home/sebastiaan/repositories/sharry/modules/webapp/target/scala-2.13/resource_managed/main/META-INF/resources/webjars/sharry-webapp/1.8.0-SNAPSHOT/img/icon-dark.svg
[info] Copy /home/sebastiaan/repositories/sharry/modules/webapp/src/main/webjar/img/logo-dark.png -> /home/sebastiaan/repositories/sharry/modules/webapp/target/scala-2.13/resource_managed/main/META-INF/resources/webjars/sharry-webapp/1.8.0-SNAPSHOT/img/logo-dark.png
[info] Copy /home/sebastiaan/repositories/sharry/modules/webapp/src/main/webjar/img/logo.svg -> /home/sebastiaan/repositories/sharry/modules/webapp/target/scala-2.13/resource_managed/main/META-INF/resources/webjars/sharry-webapp/1.8.0-SNAPSHOT/img/logo.svg
[info] Copy /home/sebastiaan/repositories/sharry/modules/webapp/src/main/webjar/sharry.js -> /home/sebastiaan/repositories/sharry/modules/webapp/target/scala-2.13/resource_managed/main/META-INF/resources/webjars/sharry-webapp/1.8.0-SNAPSHOT/sharry.js
[info] Copy /home/sebastiaan/repositories/sharry/modules/webapp/src/main/webjar/sharry.css -> /home/sebastiaan/repositories/sharry/modules/webapp/target/scala-2.13/resource_managed/main/META-INF/resources/webjars/sharry-webapp/1.8.0-SNAPSHOT/sharry.css
[info] Copy /home/sebastiaan/repositories/sharry/modules/webapp/src/main/webjar/favicon/safari-pinned-tab.svg -> /home/sebastiaan/repositories/sharry/modules/webapp/target/scala-2.13/resource_managed/main/META-INF/resources/webjars/sharry-webapp/1.8.0-SNAPSHOT/favicon/safari-pinned-tab.svg
[info] Copy /home/sebastiaan/repositories/sharry/modules/webapp/src/main/webjar/favicon/favicon-32x32.png -> /home/sebastiaan/repositories/sharry/modules/webapp/target/scala-2.13/resource_managed/main/META-INF/resources/webjars/sharry-webapp/1.8.0-SNAPSHOT/favicon/favicon-32x32.png
[info] Copy /home/sebastiaan/repositories/sharry/modules/webapp/src/main/webjar/favicon/favicon-16x16.png -> /home/sebastiaan/repositories/sharry/modules/webapp/target/scala-2.13/resource_managed/main/META-INF/resources/webjars/sharry-webapp/1.8.0-SNAPSHOT/favicon/favicon-16x16.png
[info] Copy /home/sebastiaan/repositories/sharry/modules/webapp/src/main/webjar/favicon/android-chrome-512x512.png -> /home/sebastiaan/repositories/sharry/modules/webapp/target/scala-2.13/resource_managed/main/META-INF/resources/webjars/sharry-webapp/1.8.0-SNAPSHOT/favicon/android-chrome-512x512.png
[info] Copy /home/sebastiaan/repositories/sharry/modules/webapp/src/main/webjar/favicon/android-chrome-192x192.png -> /home/sebastiaan/repositories/sharry/modules/webapp/target/scala-2.13/resource_managed/main/META-INF/resources/webjars/sharry-webapp/1.8.0-SNAPSHOT/favicon/android-chrome-192x192.png
[info] Copy /home/sebastiaan/repositories/sharry/modules/webapp/src/main/webjar/favicon/apple-touch-icon.png -> /home/sebastiaan/repositories/sharry/modules/webapp/target/scala-2.13/resource_managed/main/META-INF/resources/webjars/sharry-webapp/1.8.0-SNAPSHOT/favicon/apple-touch-icon.png
[info] Copy /home/sebastiaan/repositories/sharry/modules/webapp/src/main/webjar/favicon/mstile-150x150.png -> /home/sebastiaan/repositories/sharry/modules/webapp/target/scala-2.13/resource_managed/main/META-INF/resources/webjars/sharry-webapp/1.8.0-SNAPSHOT/favicon/mstile-150x150.png
[info] Copy /home/sebastiaan/repositories/sharry/modules/webapp/src/main/webjar/favicon/manifest.json -> /home/sebastiaan/repositories/sharry/modules/webapp/target/scala-2.13/resource_managed/main/META-INF/resources/webjars/sharry-webapp/1.8.0-SNAPSHOT/favicon/manifest.json
[info] Copy /home/sebastiaan/repositories/sharry/modules/webapp/src/main/webjar/favicon/browserconfig.xml -> /home/sebastiaan/repositories/sharry/modules/webapp/target/scala-2.13/resource_managed/main/META-INF/resources/webjars/sharry-webapp/1.8.0-SNAPSHOT/favicon/browserconfig.xml
[info] Copy /home/sebastiaan/repositories/sharry/modules/webapp/src/main/webjar/favicon/favicon.ico -> /home/sebastiaan/repositories/sharry/modules/webapp/target/scala-2.13/resource_managed/main/META-INF/resources/webjars/sharry-webapp/1.8.0-SNAPSHOT/favicon/favicon.ico
[info] compiling 20 Scala sources to /home/sebastiaan/repositories/sharry/modules/common/target/scala-2.13/classes ...
-- TYPE MISMATCH --------------- modules/webapp/src/main/elm/Page/Login/View.elm
  | => common / Compile / compileIncremental 0s
This record does not have a `logoUrlDark` field:

33|                         src flags.config.logoUrlDark
                                             ^^^^^^^^^^^
This is usually a typo. Here are the fields that are most similar:

    { logoUrl : String
    , baseUrl : String
    , iconUrl : String
    , footerText : String
    , ...
    }

So maybe logoUrlDark should be logoUrl?

                                                          Page.Login.View  ↑    
====o======================================================================o====
    ↓  Page.Register.View


-- TYPE MISMATCH ------------ modules/webapp/src/main/elm/Page/Register/View.elm

This record does not have a `logoUrlDark` field:

30|                         src flags.config.logoUrlDark
                                             ^^^^^^^^^^^
This is usually a typo. Here are the fields that are most similar:

    { logoUrl : String
    , baseUrl : String
    , iconUrl : String
    , footerText : String
    , ...
    }

So maybe logoUrlDark should be logoUrl?

[info] Running scalafix on 20 Scala sources
[info] compiling 18 Scala sources to /home/sebastiaan/repositories/sharry/modules/store/target/scala-2.13/classes ...
[info] compiling 31 Scala sources to /home/sebastiaan/repositories/sharry/modules/restapi/target/scala-2.13/classes ...
[info] Styles built
[info] Running scalafix on 18 Scala sources
[info] compiling 51 Scala sources to /home/sebastiaan/repositories/sharry/modules/backend/target/scala-2.13/classes ...
[info] Running scalafix on 51 Scala sources
[error] java.lang.RuntimeException: Nonzero exit value: 1
[error] 	at scala.sys.package$.error(package.scala:30)
[error] 	at scala.sys.process.ProcessBuilderImpl$AbstractBuilder.slurp(ProcessBuilderImpl.scala:138)
[error] 	at scala.sys.process.ProcessBuilderImpl$AbstractBuilder.$bang$bang(ProcessBuilderImpl.scala:108)
[error] 	at $c883dfcf74a7f0fafd9c$.compileElm(build.sbt:379)
[error] 	at $c883dfcf74a7f0fafd9c$.$anonfun$elmSettings$2(build.sbt:48)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] 	at sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] 	at sbt.Execute.work(Execute.scala:291)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error] 	at java.lang.Thread.run(Thread.java:748)
[error] (webapp / Compile / managedResources) Nonzero exit value: 1
[error] Total time: 38 s, completed 18-May-2021 23:20:57
```

Are they familiar to you and would you know how to fix them?

**Comment by Stamper-repo** [comment]
Note to self:
- Authentication failures for logging in are logged
- Authentication failures for using wrong password on share is missing

File: modules/restserver/src/main/scala/sharry/restserver/routes/LoginRoutes.scala (hunk 1 above)
Comment (eikek) [reply modules/restserver/src/main/scala/sharry/restserver/routes/LoginRoutes.scala:151]: Thanks for the heads up! Forget the ClientRequestInfo then! :-) Of course, this fits! And  nice examples and docs, thank you so much,

**Comment by eikek** [comment]
> Are they familiar to you and would you know how to fix them?

Yes, the ui has changed and also the openapi-spec. This requires to generate some code (client and server to be in sync) and the `sbt fix` command doesn't run this task. To quickly generate it, run `sbt openapiCodegen` - this should generate required files and then everything should compile again.

**Comment by Stamper-repo** [comment]
@eikek Proposal is finished! Hope it fits your expectations. I believe I got all brute-force-sensitive routes on the rest server. Do you know of other authentication related routes that we could protect with fail2ban?

Thanks! :D

**Comment by Stamper-repo** [comment]
(Btw, new UI looks really great!!)

**Comment by eikek** [comment]
> Do you know of other authentication related routes that we could protect with fail2ban?

No, I think these are all.

Even accounted for guessing share ids, nice. Thank you

**Pull request merged**